https://jenkins.ceph.com/job/ceph-windows-pull-requests/83662/consoleFull
```
/home/ubuntu/ceph/src/test/libcephfs/test.cc:5173:24: error: use of undeclared identifier 'FSCRYPT_MAXIO_SIZE'
    READ_WRITE_LIMIT = FSCRYPT_MAXIO_SIZE;
```

--- a/src/test/client/syncio.cc
+++ b/src/test/client/syncio.cc
@@ -23,6 +23,11 @@
 
 #include "test/client/TestClient.h"
 
+// fscrypt is supported only on linux
+#ifndef FSCRYPT_MAXIO_SIZE
+#define FSCRYPT_MAXIO_SIZE (INT_MAX & ~4095)
+#endif
+
 TEST_F(TestClient, LlreadvLlwritevInvalidFileHandleSync) {
     /* Test provding null or invalid file handle returns an error
     as expected*/
@@ -145,4 +150,76 @@ TEST_F(TestClient, LlreadvLlwritevLargeBuffersSync) {
 
   client->ll_release(fh);
   ASSERT_EQ(0, client->ll_unlink(root, filename, myperm));
+}
+
+TEST_F(TestClient, SyncIOLlreadvLlwritevSizeOutOfBound) {
+  /* sync/async i/o system calls can transfer at most INT_MAX 
+  (that is 2147483647 bytes which is ~2GiB) bytes. Therefore, 
+  test that sync'd read/write calls returned bytes do not
+  exceed this limit.*/
+  uint64_t READ_WRITE_LIMIT = INT_MAX;
+#if defined(__linux__)
+  if (fscrypt_enabled) {
+    READ_WRITE_LIMIT = FSCRYPT_MAXIO_SIZE;
+  }
+#endif
+
+  Inode *root = NULL;
+  root = client->get_root();
+  ASSERT_NE(root, (Inode *)NULL);
+
+  struct statvfs stbuf;
+  ASSERT_EQ(client->ll_statfs(root, &stbuf, myperm), 0);
+  // ensure there is enough space
+  ASSERT_GT(stbuf.f_bfree * stbuf.f_bsize, READ_WRITE_LIMIT);
+
+  int mypid = getpid();
+  char fname[256];
+  sprintf(fname, "test_synciollreadvllwritevsizeoutofbound%u", mypid);
+  Inode *file = NULL;
+  Fh *fh = NULL;
+  struct ceph_statx stx;
+  ASSERT_EQ(0, client->ll_createx(root, fname, 0666,
+      O_RDWR | O_CREAT | O_TRUNC, &file, &fh, &stx, 0, 0, myperm));
+
+  int64_t bytes_written = 0, bytes_read = 0;
+
+  const size_t BUFSIZE = 1024 * 1024 * 1024;
+  auto out_buf_0 = std::make_unique<char[]>(BUFSIZE);
+  memset(out_buf_0.get(), 'a', BUFSIZE);
+  auto out_buf_1 = std::make_unique<char[]>(BUFSIZE);
+  memset(out_buf_1.get(), 'b', BUFSIZE);
+  auto out_buf_2 = std::make_unique<char[]>(BUFSIZE);
+  memset(out_buf_2.get(), 'c', BUFSIZE);
+
+  struct iovec iov_out_over_limit[3] = {
+      {out_buf_0.get(), BUFSIZE},
+      {out_buf_1.get(), BUFSIZE},
+      {out_buf_2.get(), BUFSIZE}
+  };
+
+  auto in_buf_0 = std::make_unique<char[]>(BUFSIZE);
+  auto in_buf_1 = std::make_unique<char[]>(BUFSIZE);
+  auto in_buf_2 = std::make_unique<char[]>(BUFSIZE);
+
+  struct iovec iov_in_over_limit[3] = {
+    {in_buf_0.get(), BUFSIZE},
+    {in_buf_1.get(), BUFSIZE},
+    {in_buf_2.get(), BUFSIZE}
+  };
+
+  bytes_written = client->ll_writev(fh, iov_out_over_limit, 3, 0);
+  ASSERT_EQ(bytes_written, READ_WRITE_LIMIT);
+  ASSERT_EQ(client->ll_fsync(fh, false), 0);
+
+  // make sure the bytes_written corresponds with the file size
+  struct ceph_statx file_stx;
+  ASSERT_EQ(client->ll_getattrx(file, &file_stx, CEPH_STATX_SIZE, 0, myperm), 0);
+  ASSERT_EQ(file_stx.stx_size, READ_WRITE_LIMIT);
+
+  bytes_read = client->ll_readv(fh, iov_in_over_limit, 3, 0);
+  ASSERT_EQ(bytes_read, READ_WRITE_LIMIT);
+
+  client->ll_release(fh);
+  ASSERT_EQ(client->ll_unlink(root, fname, myperm), 0);
 }

--- a/src/test/libcephfs/CMakeLists.txt
+++ b/src/test/libcephfs/CMakeLists.txt
@@ -18,6 +18,7 @@
 
 if(WITH_LIBCEPHFS)
   add_executable(ceph_test_libcephfs
+    fscrypt_conf.cc
     test.cc
     readdir_r_cb.cc
     caps.cc
@@ -179,6 +180,7 @@ if(WITH_LIBCEPHFS)
   
   add_executable(ceph_test_libcephfs_fscrypt
     fscrypt.cc
+    fscrypt_conf.cc
     test.cc
   )
   target_link_libraries(ceph_test_libcephfs_fscrypt

--- a/src/test/libcephfs/fscrypt.cc
+++ b/src/test/libcephfs/fscrypt.cc
@@ -34,6 +34,7 @@
 #include "include/fs_types.h"
 
 #include "client/FSCrypt.h"
+#include "fscrypt_conf.h"
 
 #ifdef __linux__
 #include <limits.h>
@@ -1241,6 +1242,8 @@ int main(int argc, char **argv)
   if (r < 0)
     exit(1);
 #endif
+  // toggle the bool only after we are ready to run tests
+  fscrypt_enabled = true;
   r = RUN_ALL_TESTS();
 
   rados_shutdown(cluster);

--- a/src/test/libcephfs/fscrypt_conf.cc
+++ b/src/test/libcephfs/fscrypt_conf.cc
@@ -1,0 +1,15 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) IBM Corporation 2026
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+bool fscrypt_enabled = false;

--- a/src/test/libcephfs/fscrypt_conf.h
+++ b/src/test/libcephfs/fscrypt_conf.h
@@ -1,0 +1,17 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) IBM Corporation 2026
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#pragma once
+
+extern bool fscrypt_enabled;

--- a/src/test/libcephfs/test.cc
+++ b/src/test/libcephfs/test.cc
@@ -18,6 +18,8 @@
 #include "gtest/gtest.h"
 #include "include/cephfs/libcephfs.h"
 #include "include/ceph_fs.h"
+#include "client/FSCrypt.h"
+#include "fscrypt_conf.h"
 #include "mds/mdstypes.h"
 #include "include/stat.h"
 #include <errno.h>
@@ -50,6 +52,11 @@
 // Darwin or windows fails to define this
 #ifndef O_RSYNC
 #define O_RSYNC 0x0
+#endif
+
+// fscrypt is supported only on linux
+#ifndef FSCRYPT_MAXIO_SIZE
+#define FSCRYPT_MAXIO_SIZE (INT_MAX & ~4095)
 #endif
 
 using namespace std;
@@ -5163,4 +5170,34 @@ TEST(LibCephFS, ZeroSizeBufferAsyncReadFsync) {
   ASSERT_EQ(0, ceph_unmount(cmount));
   ceph_release(cmount);
   ceph_userperm_destroy(perms);
+}
+
+TEST(LibCephFS, BuffersOverLimit) {
+  uint64_t READ_WRITE_LIMIT = INT_MAX;
+#if defined(__linux__)
+  if (fscrypt_enabled) {
+    READ_WRITE_LIMIT = FSCRYPT_MAXIO_SIZE;
+  }
+#endif
+
+  struct ceph_mount_info *cmount;
+  ASSERT_EQ(0, ceph_create(&cmount, NULL));
+  ASSERT_EQ(0, ceph_conf_read_file(cmount, NULL));
+  ASSERT_EQ(0, ceph_conf_parse_env(cmount, NULL));
+  ASSERT_EQ(0, do_ceph_mount(cmount, "/"));
+
+  char test_file[1024];
+  sprintf(test_file, "test_buffers_over_limit_%d", getpid());
+  int fd = ceph_open(cmount, test_file, O_CREAT|O_RDWR, 0666);
+  ASSERT_LT(0, fd);
+
+  const size_t BUFSIZE = static_cast<size_t>(3) * 1024 * 1024 * 1024;
+  auto out_buf = std::make_unique<char[]>(BUFSIZE); 
+  memset(out_buf.get(), 'a', BUFSIZE);
+  auto in_buf = std::make_unique<char[]>(BUFSIZE);
+  ASSERT_EQ(ceph_write(cmount, fd, out_buf.get(), BUFSIZE, 0), READ_WRITE_LIMIT); 
+  ASSERT_EQ(ceph_read(cmount, fd, in_buf.get(), BUFSIZE, 0), READ_WRITE_LIMIT); 
+  ASSERT_EQ(0, ceph_close(cmount, fd));
+
+  ceph_shutdown(cmount);
 }

